### PR TITLE
Add docs on CodeQL Design Patterns

### PR DESCRIPTION
**Pull request by RasmusWL**

### CODE REVIEW

File: docs/ql-design-patterns.md
Comment (max-schaefer) [inline docs/ql-design-patterns.md:78]: It's perhaps worth calling out the canonical example where we do _not_ want this: `DataFlow::Configuration` and its variants are abstract, but we generally do not want to import all configurations into the same scope at once.

Comment (RasmusWL) [reply docs/ql-design-patterns.md:78]: Great idea. I ended up basically pasting your suggestion as an extra paragraph :+1:

**Comment by RasmusWL** [comment]
Requested reviews from you @tausbn and @p0 since you participated in our internal discussion about this. Not sure if we need approval from every language team before getting this merge? (that's mostly for you to answer @p0)

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:8]:
```suggestion

This pattern should be used when you want to model a user-extensible set of values ("extensibility"), while allowing restrictive subclasses, typically for the purposes of overriding predicates ("refinement"). Using a simple `abstract` class gives you the former, but makes it impossible to create overriding methods for all contributing extensions at once. Using a non-`abstract` class provides refinement-based overriding, but requires the original class to range over a closed, non-extensible set.
```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:12]:
```suggestion
Using a single `abstract` class looks like this:
```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:25]:
```suggestion
While this allows users of the library to add new types of `MySpecialExpr` (like, in this case, `ConcreteSubclass), there is no way to override the implementations of `memberPredicate` of all extensions at once.

Applying the `::Range` pattern yields the following:
```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:21]:
```suggestion

```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:60]:
```suggestion
Now, a concrete subclass can derive from `MySpecialExpr::Range` if it wants to extend the set of values in `MySpecialExpr`, and it will be required to implement the abstract `memberPredicate()`. Conversely, if it wants to refine `MySpecialExpr` and override `memberPredicate` for all extensions, it can do so by deriving from `MySpecialExpr` directly.

The key element of the pattern is to provide a field of type `MySpecialExpr::Range`, equating it to `this` in the characteristic predicate of `MySpecialExpr`. In member predicates, we can use either `this` or `range`, depending on which type has the API we need.
```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:64]:
```suggestion
Note that in some libraries, the `range` field is in fact called `self`. While the name of the field does not matter, using `range` may be less confusing in contexts like Python analysis, where `self` is by strong convention how you refer to the receiver of a method -- in QL, we denote that with the `this` keyword.
```

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:75]:
```suggestion
```

In the extra motivation I added above, I observe that you can't override the definitions of a member predicate contributed by extensions in a refining subclass that's implemented like this. Given that, this paragraph feels unnecessary. What do you think?

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:82]:
```suggestion
## Abstract classes as open or closed unions

A class declared as `abstract` in QL represents a union of its direct subtypes (restricted by the intersections of its supertypes and subject to its characteristic predicate). Depending on context, we may want this union to be considered "open" or "closed".

An open union is generally used for extensibility. For example, the abstract classes suggested by the `::Range` design pattern are explicitly intended as extension hooks. As another example, the `DataFlow::Configuration` design pattern provides an abstract class that is intended to be subclassed as a configuration mechanism.

A closed union is a class for which we do not expect users of the library to add more values. Historically, we have occasionally modelled this as `abstract` classes in QL, but these days that would be considered an anti-pattern: Abstract classes that are intended to be closed behave in surprising ways when subclassed by library users, and importing libraries that include derived classes can invalidate compilation caches and subvert the meaning of the program.

Instead, in modern QL we prefer to model closed unions as dbscheme unions, type union aliases (for unions of `newtype` branches), or via explicit `this instanceof Type1 or this instanceof Type 2 or ...` in the characteristic predicate of the closed-union class. Alternatively, we may represent a closed union as an `abstract` class similarly to the `::Range` pattern, but making the `::Range` type private to our library.
```

**Review by p0 — COMMENTED** [review]
What do you think of this batch of suggestions?

File: docs/ql-design-patterns.md
Comment (RasmusWL) [reply docs/ql-design-patterns.md:64]: I'm not 100% on board with your suggested change. I would like to put more emphasis on `range` being the recommended name for the field. How about this adapted version?

```suggestion
Note that in some libraries, the `range` field is in fact called `self`. While we do recommend using `range` for consistency, the name of the field does not matter (and using `range` avoids confusion in contexts like Python analysis that has strong usage of `self`).
```

Coming back to this, I'm realizing that the audience of this document is probably fairly invested in the codeql repo anyways, so I believe it's unlikely anyone will be confused about `self`/`this. I removed that part completely at some point, but added a small note about it again.

File: docs/ql-design-patterns.md
Comment (RasmusWL) [reply docs/ql-design-patterns.md:75]: I think if you're trying to wrap your head around this, I think this explanation is quite useful, so I would like to keep it.

File: docs/ql-design-patterns.md
Comment (RasmusWL) [reply docs/ql-design-patterns.md:81]: I might be misunderstanding, but it seems like we're addressing orthogonal problems.

I'm trying to highlight that when you are defining an open union such as an abstract class following the `::Range` pattern, you should ensure that all subclasses are imported (unless you have good reason not to). As a motivating example, when defining our `RemoteFlowSource::Range` in the Python rewrite, we _could_ have left it up to the queries to import the relevant ql modules that provide concrete subclasses of `RemoteFlowSource::Range` -- but that is an anti-pattern, and we should ensure that all concrete subclasses are imported by default ([which we do](
https://github.com/github/codeql/blob/76fd710946e5aa336721551517b0cff8daa21aff/python/ql/src/semmle/python/dataflow/new/RemoteFlowSources.qll#L4)). 

I don't quite understand the part about the closed unions, and that seems to be a different point deserving it's own heading. Can you maybe give an example of how the anti-pattern would look like? and what a "preferred" solution would be?

File: docs/ql-design-patterns.md
Comment (p0) [inline docs/ql-design-patterns.md:23]:
```suggestion
While this allows users of the library to add new types of `MySpecialExpr` (like, in this case, `ConcreteSubclass`), there is no way to override the implementations of `memberPredicate` of all extensions at once.
```

File: docs/ql-design-patterns.md
Comment (p0) [reply docs/ql-design-patterns.md:81]: I'm uncomfortable with this section as it stands, because it doesn't seem like good generic advice. It says "you should import all subtypes of an abstract type, unless you have a reason not to" -- but that's somewhat tautological. And even if we were to view it as restricted to just the `::Range` pattern, the very point of `::Range` is that end users can create new subtypes to extend the range *without modifying the core library*, and this by definition creates a situation where some subtypes are not imported by default.

My suggestion attempts to provide generic advice on when to use an abstract class (pretty much only if you are modelling an "open union", where you are anticipating that some clients want to contribute to the union). You're right that an example might help; does the following make it more understandable?

-------
As an example, suppose we want to define a `BinaryExpr` class, which has subtypes of `PlusExpr`, `MinusExpr`, and so on. Morally, this represents a closed union: We do not anticipate new kinds of `BinaryExpr` being added. Therefore, it would be undesirable to model it as an abstract class:

```ql
/** ANTI-PATTERN */
abstract class BinaryExpr extends Expr {
  Expr getLhs() { result = this.getChild(0) }
  Expr getRight() { result = this.getChild(1) }
}

class PlusExpr extends BinaryExpr {}
class MinusExpr extends BinaryExpr {}
...
```

Instead, the `BinaryExpr` class should be non-`abstract`, and we have the following options for specifying its extent:

- Define a dbscheme type `@binary_expr = @plus_expr | @minus_expr | ...` and add it as an additional super-class for `BinaryExpr`.
- Define a type alias `class RawBinaryExpr = @plus_expr | @minus_expr | ...` and add it as an additional super-class for `BinaryExpr`.
- Add a characteristic predicate of `BinaryExpr() { this instanceof PlusExpr or this instanceof MinusExpr or ... }`.

Intuitively, in a closed union, the union type is aware of all of its subtypes -- this is not the case for open unions, as in the `::Range` pattern.

------------

Does that make sense?

I think the lesson you are trying to convey perhaps deserves its own heading, and a more general phrasing. It is not specific to abstract classes. The guidance should be something like: "Importing new files can modify the behaviour of the standard library, by introducing new subtypes of `abstract` classes, by introducing new multiple inheritance relationships, or by overriding predicates. This can change query results and force evaluator cache misses."

**Review by p0 — COMMENTED** [review]
Thanks for the improvements. I think we just need to agree on what to do with the current "importing all subtypes of abstract base class" section.

File: docs/ql-design-patterns.md
Comment (RasmusWL) [reply docs/ql-design-patterns.md:63]: @p0 you marked this as resolved, but didn't apply my suggestion. Your thumbs up makes me believe you also wanted the change I suggested?

Comment (p0) [reply docs/ql-design-patterns.md]: Thanks for spotting that -- I guess I was assuming you'd click 'commit' 👍

**Comment by RasmusWL** [comment]
I'll admit that I needed a bit of time to fully process your feedback, and then I postponed fixing up things again and again. I have incorporated your suggested changes :+1: 

> I'm uncomfortable with this section as it stands, because it doesn't seem like good generic advice. It says "you should import all subtypes of an abstract type, unless you have a reason not to" -- but that's somewhat tautological. And even if we were to view it as restricted to just the `::Range` pattern, the very point of `::Range` is that end users can create new subtypes to extend the range without modifying the core library, and this by definition creates a situation where some subtypes are not imported by default.

I'm interpreting your _that's somewhat tautological_ to mean that what I'm saying is obviously true. But I firmly believe that it's important. I recall my more inexperienced days of CodeQL development, where I didn't fully understand the impact of having a file imported or not -- and the main point I'm trying to get across is that this is actually important to get right.

**Review by p0 — APPROVED** [review]
Thanks for all the fixes -- this LGTM now.

**Pull request merged**
